sys/senml/senml.c

```diff
@@ -0,0 +1,416 @@
+        chars_written = snprintf(&output[ins_pos], buf_len, "\"bv\":%f,",
+                                 pack->base_info->base_value);
+
+        if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+            return -1;
+        }
+
+        ins_pos += chars_written;
+        buf_len -= chars_written;
+    }
+
+encode_records:
+
+    ins_pos_after_bi = ins_pos;
+
+    if (pack->num == 0) {
+        // this means no base info and no records
+        if (ins_pos == 1) {
+            output[ins_pos] = ']';
+            ins_pos++;
+            buf_len--;
+            output[ins_pos] = '\0';
+            ins_pos++;
+            buf_len--;
+            return 0;
+        }
+        // this means there is base info but records, which is illegal
+        else {
+            return -1;
+        }
+    }
+
+    if (buf_len < 4) {   // we now need space for at least '}', ']', '\0' and perhaps '{' before
+        return -1;
+    }
+
+    buf_len -= 1;   // reserve space for '\0' (trailing comma will be overwritten with ']')
+
+    for (size_t i = 0; i < pack->num; i++) {
+        senml_record_t *curr_record = &(pack->records[i]);
+
+        size_t buf_len_old;
+
+        if (i != 0 || ins_pos == 1) {
+            output[ins_pos] = '{';
+            ins_pos++;
+            buf_len--;
+        }
+
+        buf_len_old = buf_len;
+
+        if (curr_record->name) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"n\":\"%s\",",
+                                     curr_record->name);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->unit) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"u\":\"%s\",",
+                                     curr_record->unit);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->time != 0) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"t\":%f,",
+                                     curr_record->time);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->update_time != 0) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"ut\":%f,",
+                                     curr_record->update_time);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->value_sum != 0) {
```

Comment:
NaN would be nice, but it is safe to use on all platforms? This would also add complexity to the creation of empty records, because `memset(&record, 0, sizeof(record))` would not be sufficient anymore